### PR TITLE
Bulk Batch creation

### PR DESCRIPTION
**Pull request by patins1**
This pull requests optimizes Batch creation for ArrayDataset when using StackBatchifier .
My DJL application learns the data now in 58550ms  rather than 79843ms so 36% faster!
For applications not using range-based indexing but indices-based indexing, my DJL application would run in 66341ms , so still 20% faster. This is true for PyTorch, i get a test failure for MxNet, so this solution maybe restricted to PyTorch

### CODE REVIEW

**Comment by patins1** [comment]
The problems with MxNet is fixed by using "take" operation instead of "pick".
TrainMnist executes now 4 times faster for MxNet .

**Comment by lanking520** [comment]
@KexinFeng can you help to check on the NDIndex part changes?

**Comment by KexinFeng** [comment]
@patins1 It looks like `BulkDataIterable` is not covered by any unit test. You mentioned that using `BulkDataIterable` is more efficient. Could you add a unit test that covers this class you added?

**Comment by KexinFeng** [comment]
> The problems with MxNet is fixed by using "take" operation instead of "pick".
> TrainMnist executes now 4 times faster for MxNet .

Could you integrate it with the existing `take` function in [PR](https://github.com/deepjavalibrary/djl/pull/1649)?

File: api/src/main/java/ai/djl/ndarray/index/full/NDIndexFullPick.java [hunk 1]
```diff
@@ -72,6 +73,7 @@ public static Optional<NDIndexFullPick> fromIndex(NDIndex index, Shape target) {
                             "Only rank-1 indexing array is supported for pick");
                 }
                 fullPick = new NDIndexFullPick(indexElem, axis);
+                fullPick.setIndexTake(true);
```
Comment (KexinFeng) [inline api/src/main/java/ai/djl/ndarray/index/full/NDIndexFullPick.java:76]: Here, I tried to use take for MXNet engine too see https://github.com/deepjavalibrary/djl/pull/1802, and the PR tests all passed. But it caused the https://github.com/deepjavalibrary/djl/issues/1800. Here in your edition, setting fullPick.setIndexTake(true); will trigger this problem again.

**Review by KexinFeng — CHANGES_REQUESTED** [review]
If you could add a unit test that covers `BulkDataIterable` and that shows the necessity of using `take` on MXNet engine, it'd be great.

File: api/src/main/java/ai/djl/ndarray/index/full/NDIndexFullPick.java (hunk 1 above)
Comment (patins1) [reply api/src/main/java/ai/djl/ndarray/index/full/NDIndexFullPick.java]: `NDArray.take(NDArray)` returns a shape that is not useful for the functionality of this pull request:

`Returns a partial NDArray pointed by index according to linear indexing, and the output is of the same shape as index.`

Given `ndArray` of shape e.g. `(a,b,c,d)` and `index` of shape `(e) `, then I assumed initially that `ndArray.take(index)` would return an array of shape` (e,b,c,d)` .  Therefore I restyled my initial implementation, using NDIndexFullTake .  If anybody has a better idea, let me know, at least the test cases go green now.

**Comment by KexinFeng** [comment]
@patins1 Thanks for dealing with this issue in such a timely manner! But it looks like the unittest that covers the new file api/src/main/java/ai/djl/training/dataset/BulkDataIterable.java is still missing. You mentioned you have tested it, which display efficiency increase. Could you add it into the proper unit test file? 

Also you mentioned 
> The problems with MxNet is fixed by using "take" operation instead of "pick".
> TrainMnist executes now 4 times faster for MxNet .

It'd be better to have unit test for this too, which will prevent future edition from breaking this fix.

Thanks!

**Comment by patins1** [comment]
Tests added

**Comment by KexinFeng** [comment]
@siddvenk Hi Siddarth, in this pr, we have changed the definition of get(NDArray index) from pick to take, and have given the warning. This will affect the result you mentioned in https://github.com/deepjavalibrary/djl/issues/1800. See the test in commit "add index test and clean code" 8ffac00e6584c96500e6ac96ba35cc45513f1f4e. Pick can still be used though by addPickDim().
This change is for the purpose of making it consistent with numpy and pytorch engine, also here it is mor efficient.
Should we update the relevant part mentioned in https://github.com/deepjavalibrary/djl/issues/1800?

**Review by KexinFeng — APPROVED** [review]

**Review by zachgk — APPROVED** [review]
There is one small problem I noticed, but otherwise it looks good. I think we should just merge it for now and we can fix it in a later PR

**Comment by KexinFeng** [comment]
> @siddvenk Hi Siddarth, in this pr, we have changed the definition of get(NDArray index) from pick to take, and have given the warning. This will affect the result you mentioned in #1800. See the test in commit "add index test and clean code" [8ffac00](https://github.com/deepjavalibrary/djl/commit/8ffac00e6584c96500e6ac96ba35cc45513f1f4e). Pick can still be used though by addPickDim(). This change is for the purpose of making it consistent with numpy and pytorch engine, also it is more efficient as shown here. Should we update the relevant part of the _Dive into Deep Learning for Java_ book mentioned in #1800?

**Comment by KexinFeng** [comment]
> @siddvenk Hi Siddarth, in this pr, we have changed the definition of get(NDArray index) from pick to take, and have given the warning. This will affect the result you mentioned in #1800. See the test in commit "add index test and clean code" [8ffac00](https://github.com/deepjavalibrary/djl/commit/8ffac00e6584c96500e6ac96ba35cc45513f1f4e). Pick can still be used though by addPickDim(). This change is for the purpose of making it consistent with numpy and pytorch engine, also it is more efficient as shown here. Should we update the relevant part of the _Dive into Deep Learning for Java_ book mentioned in #1800?

**Pull request merged**